velox/exec/fuzzer/JoinMaker.h

```diff
@@ -0,0 +1,271 @@
+  enum JoinOrder { NATURAL, FLIPPED };
+
+  struct PlanWithSplits {
+    core::PlanNodePtr plan;
+    core::PlanNodeId probeScanId;
+    core::PlanNodeId buildScanId;
+    std::unordered_map<core::PlanNodeId, std::vector<velox::exec::Split>>
+        splits;
+    core::ExecutionStrategy executionStrategy{
+        core::ExecutionStrategy::kUngrouped};
+    int32_t numGroups;
+
+    explicit PlanWithSplits(
+        const core::PlanNodePtr& _plan,
+        const core::PlanNodeId& _probeScanId = "",
+        const core::PlanNodeId& _buildScanId = "",
+        const std::unordered_map<
+            core::PlanNodeId,
+            std::vector<velox::exec::Split>>& _splits = {},
+        core::ExecutionStrategy _executionStrategy =
+            core::ExecutionStrategy::kUngrouped,
+        int32_t _numGroups = 0)
+        : plan(_plan),
+          probeScanId(_probeScanId),
+          buildScanId(_buildScanId),
+          splits(_splits),
+          executionStrategy(_executionStrategy),
+          numGroups(_numGroups) {}
+  };
+
+  PlanWithSplits makeHashJoin(
+      const InputType inputType,
+      const PartitionStrategy partitionStrategy,
+      const JoinOrder joinOrder) const;
+
+  PlanWithSplits makeHashJoinWithTableScan(
+      const std::optional<int32_t> numGroups,
+      const JoinOrder joinOrder) const;
+
+  PlanWithSplits makeMergeJoin(
+      const InputType inputType,
+      const JoinOrder joinOrder) const;
+
+  PlanWithSplits makeMergeJoinWithTableScan(const JoinOrder joinOrder) const;
+
+  PlanWithSplits makeNestedLoopJoin(
+      const InputType inputType,
+      const JoinOrder joinOrder) const;
+
+  PlanWithSplits makeNestedLoopJoinWithTableScan(
+      const JoinOrder joinOrder) const;
+
+  /// Returns whether or not the join type supports reversing the order of the
+  /// inputs in a hash join.
+  bool supportsFlippingHashJoin() const;
+
+  /// Returns whether or not the join type supports reversing the order of the
+  /// inputs in a merge join.
+  bool supportsFlippingMergeJoin() const;
+
+  /// Returns whether or not the join type supports reversing the order of the
+  /// inputs in a nested loop join.
+  bool supportsFlippingNestedLoopJoin() const;
+
+  bool supportsMergeJoin() const {
+    return core::MergeJoinNode::isSupported(joinType_);
+  }
+
+  bool supportsNestedLoopJoin() const {
+    return core::NestedLoopJoinNode::isSupported(joinType_);
+  }
+
+  /// Returns whether or not the types of the sources allow them to be read as
+  /// splits using a TableScan.
+  bool supportsTableScan() const;
+
+ private:
+  // Get the sources in the order they should be read according to `joinOrder`.
+  std::vector<std::shared_ptr<JoinSource>> getJoinSources(
+      const JoinOrder joinOrder) const;
+
+  // Get the JoinType to use given `joinOrder`. Throws if the joinType does not
+  // support the given JoinOrder.
+  core::JoinType getJoinType(const JoinOrder joinOrder) const;
+
+  std::string makeNestedLoopJoinFilter(
```

Comment:
nit: Can we rename these private helper method to be makeNestedLoopJoinFilter**Impl**(), etc. to differentiate them from the public APIs?